server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -1062,12 +1066,20 @@ public void publish(
                 assert getLocalNode().equals(clusterState.getNodes().get(getLocalNode().getId())) :
                     getLocalNode() + " should be in published " + clusterState;
 
+                final long publicationContextConstructionStartMillis = transportService.getThreadPool().relativeTimeInMillis();
                 final PublicationTransportHandler.PublicationContext publicationContext =
                     publicationHandler.newPublicationContext(clusterStatePublicationEvent);
+                clusterStatePublicationEvent.setPublicationContextConstructionElapsedMillis(
+                    transportService.getThreadPool().relativeTimeInMillis() - publicationContextConstructionStartMillis);
```

Comment:
Is there a concern that if this is < 200ms that it won't be counted with the use of the `CachedTimeThread` that only updates the clock every 200ms? Should this use "real" time instead?

Reply:
My thinking was that if these things are only taking 200ms or so then we probably don't care, or more precisely we only care if they're happening with high frequency and in the high frequency case we will pick up the occasional 200ms clock tick with a probability that matches the mean duration, so on balance it'll all work out.

Also it's a pain to write tests that assert things about time when using `System::nanoTime`. I can fix up the assertions that say that at least `T` ms elapsed, but I'd just have to drop the ones with an upper bound. Maybe it's not that bad, the only one that I can't easily add an artificial delay to is the publication context construction.

I'll try out some ideas next week. Do you think we should accumulate the timers in nanoseconds too? `Long.MAX_VALUE` nanoseconds is ~300 years, I don't think there's much risk of overflow.

Reply:
Ok in 9b0c5f3 I added a method to expose the raw (uncached) relative time to `ThreadPool` which we can still override to solve the testing pain, but which has better granularity. I stuck with milliseconds.